NullPointerException thrown for UPConfig.getGroups()

### Before reporting an issue

- [X] I have read and understood the above terms for submitting issues, and I understand that my issue may be closed without action if I do not follow them.

### Area

admin/ui

### Describe the bug

```sh
ERROR [org.keycloak.services.error.KeycloakErrorHandler] (executor-thread-7) Uncaught server error: java.lang.NullPointerException: Cannot invoke "org.keycloak.representations.userprofile.config.UPConfig.getGroups()" because "config" is null
	at org.keycloak.userprofile.config.UPConfigUtils.validateAttributes(UPConfigUtils.java:112)
	at org.keycloak.userprofile.config.UPConfigUtils.validate(UPConfigUtils.java:95)
	at org.keycloak.userprofile.DeclarativeUserProfileProvider.lambda$createUserDefinedProfileDecorator$8(DeclarativeUserProfileProvider.java:523)
	at java.base/java.util.concurrent.ConcurrentHashMap.computeIfAbsent(ConcurrentHashMap.java:1708)
	at org.keycloak.userprofile.DeclarativeUserProfileProvider.configureUserProfile(DeclarativeUserProfileProvider.java:184)
	at org.keycloak.userprofile.AbstractUserProfileProvider.createUserProfile(AbstractUserProfileProvider.java:328)
	at org.keycloak.userprofile.AbstractUserProfileProvider.create(AbstractUserProfileProvider.java:205)
	at org.keycloak.services.resources.admin.UserResource.getUser(UserResource.java:325)
	at org.keycloak.services.resources.admin.UserResource$quarkusrestinvoker$getUser_7e789726117d7cfd3ef1d9ddd01596d8e6add5b0.invoke(Unknown Source)
	at org.jboss.resteasy.reactive.server.handlers.InvocationHandler.handle(InvocationHandler.java:29)
	at io.quarkus.resteasy.reactive.server.runtime.QuarkusResteasyReactiveRequestContext.invokeHandler(QuarkusResteasyReactiveRequestContext.java:141)
	at org.jboss.resteasy.reactive.common.core.AbstractResteasyReactiveContext.run(AbstractResteasyReactiveContext.java:145)
	at io.quarkus.vertx.core.runtime.VertxCoreRecorder$14.runWith(VertxCoreRecorder.java:576)
	at org.jboss.threads.EnhancedQueueExecutor$Task.run(EnhancedQueueExecutor.java:2513)
	at org.jboss.threads.EnhancedQueueExecutor$ThreadBody.run(EnhancedQueueExecutor.java:1538)
	at org.jboss.threads.DelegatingRunnable.run(DelegatingRunnable.java:29)
	at org.jboss.threads.ThreadLocalResettingRunnable.run(ThreadLocalResettingRunnable.java:29)
	at io.netty.util.concurrent.FastThreadLocalRunnable.run(FastThreadLocalRunnable.java:30)
	at java.base/java.lang.Thread.run(Thread.java:840)
```

### Version

23.0.1

### Expected behavior

No errors when attempting to view User Detail pages

### Actual behavior

When trying to go into the User details for some Realms, getting a "Network response was not OK". The suggesting reload of the page does not resolve.

![image](https://github.com/keycloak/keycloak/assets/101352781/915fb077-698a-4635-bc37-2aa9d51d6815)

### How to Reproduce?

Try to navigate into a user detail when the error is present. 

### Anything else?

Doesn't appear to affect all realms so possibly related to some partial migration step?

Installation was updated automatically using the Keycloak Operator.

@polaris940 Thanks for the details.

When running migration there is an error happening due to the realm not being available from the session context:

```
Caused by: java.lang.NullPointerException: Cannot invoke "org.keycloak.models.RealmModel.getClientScopesStream()" because "realm" is null
	at org.keycloak.userprofile.config.UPConfigUtils.validateScopes(UPConfigUtils.java:207)
	at org.keycloak.userprofile.config.UPConfigUtils.validateAttribute(UPConfigUtils.java:177)
	at org.keycloak.userprofile.config.UPConfigUtils.lambda$validateAttributes$2(UPConfigUtils.java:130)
	at java.base/java.util.ArrayList.forEach(ArrayList.java:1511)
	at org.keycloak.userprofile.config.UPConfigUtils.validateAttributes(UPConfigUtils.java:130)
	at org.keycloak.userprofile.config.UPConfigUtils.validate(UPConfigUtils.java:107)
	at org.keycloak.userprofile.DeclarativeUserProfileProviderFactory.validateConfiguration(DeclarativeUserProfileProviderFactory.java:269)
	at org.keycloak.models.jpa.RealmAdapter.updateComponent(RealmAdapter.java:2088)
	at org.keycloak.models.cache.infinispan.RealmAdapter.updateComponent(RealmAdapter.java:1561)
	at org.keycloak.migration.migrators.MigrateTo23_0_0.updateUserProfileConfig(MigrateTo23_0_0.java:82)
	at org.keycloak.migration.migrators.MigrateTo23_0_0.migrateRealm(MigrateTo23_0_0.java:55)
	at java.base/java.util.stream.ForEachOps$ForEachOp$OfRef.accept(ForEachOps.java:183)
	at java.base/java.util.stream.ReferencePipeline$3$1.accept(ReferencePipeline.java:197)
	at java.base/java.util.stream.ReferencePipeline$3$1.accept(ReferencePipeline.java:197)
	at java.base/java.util.stream.ReferencePipeline$2$1.accept(ReferencePipeline.java:179)
	at java.base/java.util.stream.ReferencePipeline$3$1.accept(ReferencePipeline.java:197)
```

So the migration happened here but failed due to the above. Not sure why you did not get the same issue. I'm migrating from `22.0.4` (realm + UP enabled + single custom attribute added to config).

The error also indicates we don't have enough coverage from migration tests, I suppose.

--- a/model/legacy-private/src/main/java/org/keycloak/migration/migrators/MigrateTo23_0_0.java
+++ b/model/legacy-private/src/main/java/org/keycloak/migration/migrators/MigrateTo23_0_0.java
@@ -25,6 +25,7 @@ import org.jboss.logging.Logger;
 import org.keycloak.authentication.AuthenticationFlow;
 import org.keycloak.component.ComponentModel;
 import org.keycloak.migration.ModelVersion;
+import org.keycloak.models.KeycloakContext;
 import org.keycloak.models.KeycloakSession;
 import org.keycloak.models.RealmModel;
 import org.keycloak.representations.idm.RealmRepresentation;
@@ -43,7 +44,16 @@ public class MigrateTo23_0_0 implements Migration {
 
     @Override
     public void migrate(KeycloakSession session) {
-        session.realms().getRealmsStream().forEach(this::migrateRealm);
+        session.realms().getRealmsStream().forEach(realm -> {
+            KeycloakContext context = session.getContext();
+
+            try {
+                context.setRealm(realm);
+                migrateRealm(realm);
+            } finally {
+                context.setRealm(null);
+            }
+        });
     }
 
     @Override

--- a/server-spi-private/src/main/java/org/keycloak/models/utils/RepresentationToModel.java
+++ b/server-spi-private/src/main/java/org/keycloak/models/utils/RepresentationToModel.java
@@ -67,6 +67,7 @@ import org.keycloak.models.FederatedIdentityModel;
 import org.keycloak.models.GroupModel;
 import org.keycloak.models.IdentityProviderMapperModel;
 import org.keycloak.models.IdentityProviderModel;
+import org.keycloak.models.KeycloakContext;
 import org.keycloak.models.KeycloakSession;
 import org.keycloak.models.ModelException;
 import org.keycloak.models.ProtocolMapperModel;
@@ -121,7 +122,13 @@ public class RepresentationToModel {
 
 
     public static void importRealm(KeycloakSession session, RealmRepresentation rep, RealmModel newRealm, boolean skipUserDependent) {
-        session.getProvider(DatastoreProvider.class).getExportImportManager().importRealm(rep, newRealm, skipUserDependent);
+        KeycloakContext context = session.getContext();
+        try {
+            context.setRealm(newRealm);
+            session.getProvider(DatastoreProvider.class).getExportImportManager().importRealm(rep, newRealm, skipUserDependent);
+        } finally {
+            context.setRealm(null);
+        }
     }
 
     public static void importRoles(RolesRepresentation realmRoles, RealmModel realm) {

--- a/testsuite/integration-arquillian/tests/base/src/main/java/org/keycloak/testsuite/arquillian/containers/AbstractQuarkusDeployableContainer.java
+++ b/testsuite/integration-arquillian/tests/base/src/main/java/org/keycloak/testsuite/arquillian/containers/AbstractQuarkusDeployableContainer.java
@@ -31,6 +31,7 @@ import java.security.cert.X509Certificate;
 import java.util.ArrayList;
 import java.util.Collections;
 import java.util.HashMap;
+import java.util.Iterator;
 import java.util.List;
 import java.util.Map;
 import java.util.concurrent.TimeUnit;
@@ -59,6 +60,7 @@ import org.jboss.shrinkwrap.descriptor.api.Descriptor;
 import org.keycloak.common.crypto.FipsMode;
 import org.keycloak.testsuite.arquillian.SuiteContext;
 import org.keycloak.testsuite.model.StoreProvider;
+import org.keycloak.utils.StringUtil;
 
 public abstract class AbstractQuarkusDeployableContainer implements DeployableContainer<KeycloakQuarkusConfiguration> {
 
@@ -206,8 +208,37 @@ public abstract class AbstractQuarkusDeployableContainer implements DeployableCo
         }
 
         addStorageOptions(storeProvider, commands);
+        addFeaturesOption(commands);
 
         return commands;
+    }
+
+    protected void addFeaturesOption(List<String> commands) {
+        String defaultFeatures = configuration.getDefaultFeatures();
+
+        if (StringUtil.isBlank(defaultFeatures)) {
+            return;
+        }
+
+        if (commands.stream().anyMatch(List.of("import", "export")::contains)) {
+            return;
+        }
+
+        StringBuilder featuresOption = new StringBuilder("--features=").append(defaultFeatures);
+        Iterator<String> iterator = commands.iterator();
+
+        while (iterator.hasNext()) {
+            String command = iterator.next();
+
+            if (command.startsWith("--features")) {
+                featuresOption = new StringBuilder(command);
+                featuresOption.append(",").append(defaultFeatures);
+                iterator.remove();
+                break;
+            }
+        }
+
+        commands.add(featuresOption.toString());
     }
 
     protected List<String> configureArgs(List<String> commands) {

--- a/testsuite/integration-arquillian/tests/base/src/main/java/org/keycloak/testsuite/arquillian/containers/KeycloakQuarkusConfiguration.java
+++ b/testsuite/integration-arquillian/tests/base/src/main/java/org/keycloak/testsuite/arquillian/containers/KeycloakQuarkusConfiguration.java
@@ -47,6 +47,8 @@ public class KeycloakQuarkusConfiguration implements ContainerConfiguration {
 
     private FipsMode fipsMode = FipsMode.valueOfOption(System.getProperty("auth.server.fips.mode"));
 
+    private String defaultFeatures;
+
     @Override
     public void validate() throws ConfigurationException {
         int basePort = getBindHttpPort();
@@ -228,5 +230,13 @@ public class KeycloakQuarkusConfiguration implements ContainerConfiguration {
 
     public void setFipsMode(FipsMode fipsMode) {
         this.fipsMode = fipsMode;
+    }
+
+    public void setDefaultFeatures(String defaultFeatures) {
+        this.defaultFeatures = defaultFeatures;
+    }
+
+    public String getDefaultFeatures() {
+        return defaultFeatures;
     }
 }

--- a/testsuite/integration-arquillian/tests/base/src/main/java/org/keycloak/testsuite/arquillian/containers/KeycloakQuarkusServerDeployableContainer.java
+++ b/testsuite/integration-arquillian/tests/base/src/main/java/org/keycloak/testsuite/arquillian/containers/KeycloakQuarkusServerDeployableContainer.java
@@ -73,6 +73,7 @@ public class KeycloakQuarkusServerDeployableContainer extends AbstractQuarkusDep
         commands.add(getCommand());
         commands.add("-v");
         commands.add(command);
+        addFeaturesOption(commands);
         if (args != null) {
             commands.addAll(Arrays.asList(args));
         }

--- a/testsuite/integration-arquillian/tests/base/src/test/resources/arquillian.xml
+++ b/testsuite/integration-arquillian/tests/base/src/test/resources/arquillian.xml
@@ -649,6 +649,7 @@
             <property name="javaOpts">-Xms512m -Xmx512m -XX:MetaspaceSize=96M -XX:MaxMetaspaceSize=512m
                 -Djava.net.preferIPv4Stack=true -Dauth.server.db.host=some
             </property>
+            <property name="defaultFeatures">${auth.server.feature}</property>
         </configuration>
     </container>
 

--- a/testsuite/integration-arquillian/tests/pom.xml
+++ b/testsuite/integration-arquillian/tests/pom.xml
@@ -1534,6 +1534,8 @@
                                     <auth.server.migration>true</auth.server.migration>
                                     <keycloak.migration.home>${containers.home}/auth-server-migration</keycloak.migration.home>
                                     <migration.import.props.previous>${migration.import.props.previous}</migration.import.props.previous>
+                                    <auth.server.feature>${auth.server.feature}</auth.server.feature>
+                                    <auth.server.feature>declarative-user-profile</auth.server.feature>
                                 </systemPropertyVariables>
                             </configuration>
                         </plugin>

